metricflow/model/validations/materializations.py

```diff
@@ -0,0 +1,315 @@
+import logging
+from typing import List, Set, Union
+
+from metricflow.model.objects.elements.dimension import DimensionType
+from metricflow.model.objects.materialization import (
+    Materialization,
+    MaterializationDestination,
+    MaterializationFormat,
+    MaterializationLocation,
+)
+from metricflow.model.objects.user_configured_model import UserConfiguredModel
+from metricflow.model.semantic_model import SemanticModel
+from metricflow.model.validations.data_sources import DataSourceTimeDimensionWarningsRule
+from metricflow.model.validations.dim_names import DimensionAndIdentifierNameValidator
+from metricflow.specs import DimensionReference, DimensionSpec, MetricSpec, TimeDimensionSpec
+from metricflow.naming.linkable_spec_name import StructuredLinkableSpecName
+from metricflow.time.time_granularity_solver import (
+    RequestTimeGranularityException,
+    TimeGranularitySolver,
+    TimeGranularityValidationReference,
+)
+from metricflow.model.validations.validator_helpers import (
+    ModelValidationRule,
+    ValidationIssue,
+    ValidationError,
+    ValidationIssueType,
+    validate_safely,
+    ValidationIssueLevel,
+)
+
+logger = logging.getLogger(__name__)
+
+WILDCARD_NAME = "*"
+
+
+class ValidMaterializationRule(ModelValidationRule):
+    """Check that a materialization config is valid.
+
+    * Metrics listed are valid.
+    * Dimensions or identifiers are valid for the listed metrics.
+    * Dimensions listed in a rollup are valid.
+    * Primary time dimension is listed in the dimensions section.
+    """
+
+    # Maximum number of wildcards in the dimensions for a rollup.
+    # rollups might look something like ["ds", "country", "*", "*"]
+    MAX_WILDCARDS_IN_ROLLUP = 6
+
+    @staticmethod
+    def _find_primary_time_dimension_ref(model: UserConfiguredModel) -> DimensionReference:
+        for data_source in model.data_sources or []:
+            for dimension in data_source.dimensions or []:
+                if (
+                    dimension.type == DimensionType.TIME
+                    and dimension.type_params is not None
+                    and dimension.type_params.is_primary
+                ):
+                    return dimension.name
+        raise RuntimeError(
+            f"No primary time dimension in the model. This should have been caught by "
+            f"{DataSourceTimeDimensionWarningsRule}"
+        )
+
+    @staticmethod
+    @validate_safely(whats_being_done="checking materialization is defined correctly")
+    def _validate_materialization(
```

Comment:
Actually, this can be greatly simplified by using `MetricSemantics.element_specs_for_metrics()`. Can you take a look at the function? If you do that, you won't need `DimensionAndIdentifierNameValidator` or `TimeGranularitySolver`.